### ISSUE TITLE
Design Doc: HTTP Healthcheck on Deploy

### ISSUE THREAD

**Issue by bacongobbler** [deis/deis#4045]
refs the new proposal in #4019. All input/feedback on the doc is welcome.
refs #3813

Please note that this document is being maintained at https://gist.github.com/bacongobbler/d2f0ef2311d1e3e28997 to keep a living history of any changes made.

During a typical deployment of an application release, the controller performs a status check from Fleet after it has been deployed to ensure that it is running. However, Fleet is a glorified systemd wrapper, which means that it only checks systemd if the process is running (in this case, `docker run`). If a container's web process takes some time to boot, this may cause a false positive status check from the user's perspective. This proposal introduces a new concept into Deis known as a health check, which will make the controller check Deis' Publisher component for real-time stats on the app to ensure the container is running before recycling the old release.

Goals
-----

The main goals of this proposal is to

 - ensure a zero-downtime deployment between deployments of new releases
 - allow the user to configure the URL path, the health check delay, and the health check request timeout value.

Code Changes
------------

In order for this change to occur, there needs to be a way for users to configure both the expected health check URL as well as the initial delay and timeout value of the request. Since the `Config` object currently exists and the health checks are considered part of an application's configuration, the code changes in the controller is fairly minimal:

 - `deis config:set HEALTHCHECK_URL=/` will configure a health check URL for the application
 - the expected status code from the health check is a 200 OK. This is non-configurable.
 - `deis config:set HEALTHCHECK_PORT=5432` will configure a health check port for the application (default: 5000)
 - `deis config:set HEALTHCHECK_INITIAL_DELAY=10` will tell Deis' Publisher to wait 10 seconds before performing a health check (default: 0)
 - `deis config:set HEALTHCHECK_TIMEOUT=10` will tell Deis' Publisher that the request may take up to 10 seconds before the app is considered "down" (default: 5).

These keys will be exposed to Etcd at /Deis/config/appname/health check_URL=/ so deis's Publisher can retrieve these values to perform its job.

Once `HEALTHCHECK_URL` is set, the controller will do the typical application release workflow. However, once the controller is notified by Fleet that the new version of the application is "up", the controller will wait a minimum of x+y seconds, where x is equal to the value of HEALTHCHECK_INITIAL_DELAY, and y is equal to the value of HEALTHCHECK_TIMEOUT. By default this is 5 seconds. Once the waiting is over, the controller will then perform an API request against the Publisher to inspect the container's health (more on that below).

On the Publisher's side, before it publishes a container to Etcd, it will check Etcd for these keys. If health check_URL is set, then it will perform the health check. It will also expose an API which the controller can now ping. The API will be very simple:

Example request:

	GET /up

Example Response:

	[
		"foo_v2.web.1",
		"foo_v2.web.3",
		"bar_v3.cmd.1"
	]

This response is a list of all containers which Publisher has exposed to Etcd and are considered "up". It will be the responsibility of the controller to parse this response to check if the containers deployed are up based on the container ID. If the health check passes, the rest of the deployment workflow will then continue, recycling old containers and returning a successful response to the client. If only a few are down (few being anything less than all of the new containers), a warning will be raised, but the operation will be considered "passed". However, if health check failed, the controller will respond with an appropriate error and roll back to the old version of the application.

Issues to Note
--------------

With this approach, We have some legacy code in the Publisher which will only publish the latest version of the container at any given time (see #2413 for a history lesson). This was put in place to prevent two existing versions of the same app to be published to the router. We'll have to be extra careful not to break this functionality, but we also have to make sure that we're still publishing the older version of the controller to the load balancer (Deis-router) before the new container passes the health check. For example, if the user has set an initial delay that is higher than the TTL of the Etcd keys, this could cause the old version of the containers to come down, causing downtime. This can be mitigated with proper unit tests within Publisher to ensure that an older version of the container is still getting published at this time.

Testing
-------

In order for this proposal to work, we need to ensure a zero-downtime deployment, as well as proper errors to be returned when an application fails to respond in time or with an incorrect status code. Since the controller has a great suite for unit tests, most of the workflow for the client will be covered by unit tests to ensure that the proper errors are propagated when certain cases occur. However, the zero-downtime approach will be covered by a combination of unit tests in Deis-Publisher as well as some comprehensive integration tests. How this will be accomplished is by invoking a goroutine that will continuously ping the app URL. If a 404 response shows up (due to #3774), then the deployment failed as a zero-downtime deployment.

**Comment by gabrtv** [comment]
A few thoughts:

1. Let's take a closer look at GoogleCloudPlatform/kubernetes#365 and make sure our interface is compatible (right now it's not)
2. I understand publisher has proved a problematic place for this code, but I really don't like the idea of the controller implementing continuous anything.  What about the router?  Aren't there Nginx directives we could use to avoid the blips?
3. It's great that we're taking a test-driven approach to this.  Let's make sure we test for the exact error conditions we see today on a slow starting application (502's, 404's, etc.)

**Comment by Blystad** [comment]
Alright, so, this is a feature I am really looking forward to :+1: 

So far I have one comment:

> This health check will continuously perform an HTTP request every second for 60 seconds.

If possible, make that configurable...us people running Java applications which do classpath scanning, dependency tree building, dependency injection, database initialization and so on so forth during boot, can often take a bit more then 60 seconds to start up (especially on older monolithic applications). I would prefer a default of perhaps 2 or 5 minutes. Are there any good reasons why it should be limited to 1 minute? The routing layer won't timeout for 20 minutes.

**Comment by bacongobbler** [comment]
> Are there any good reasons why it should be limited to 1 minute?

Simply put, precedence set by Heroku. We can easily make this configurable though, no doubt :)

**Comment by Blystad** [comment]
Perfect. I was also a bit sure about this part:

>  If the health check passes, the rest of the deployment workflow will continue, recycling old containers and returning a successful response to the client. 

My original thought was that this might cause problems if an application doing database migration or something is slow (or, alternatively, connecting to database takes a few seconds too much) (and in that case, maybe allow for a configurable number of health check failures). But I just did a test with a Spring Boot application I have (fyi, Spring Boot ships with a health endpoint solution), and it seemed to handle database migrations fine, and report the correct status all the way.

**Comment by bacongobbler** [comment]
@gabrtv to address your comments:

> Let's take a closer look at GoogleCloudPlatform/kubernetes#365 and make sure our interface is compatible (right now it's not)

I just finished talking with @carmstrong about this and he said the same thing, so needless to say I'll look into it. :)

> I understand publisher has proved a problematic place for this code, but I really don't like the idea of the controller implementing continuous anything. What about the router? Aren't there Nginx directives we could use to avoid the blips?

Sadly that's an NGINX+ feature, so it's unavailable on the open-source side.

**Comment by mboersma** [comment]
> This health check will continuously perform an HTTP request every second for 60 seconds.

This also made me stop and think. Perhaps that should be the default, but the time should be configurable? And generally speaking I'm also wary of putting functionality in the controller if we can think of another way to do it.

Deis currently does a full A->B deploy, meaning we bring up X containers to replace the X currently running, which is naive and problematic if you don't have basically double the resources needed to run your app containers. It would be useful if that strategy were also configurable, so a user could ask to roll the new release out in groups of 3, for example. But I assume making the release promotion strategy configurable is out of scope for this particular change...

We have a manual "zero downtime" test script that would be a good starting point for the integration tests.

**Comment by jchauncey** [comment]
Normally your service discovery agent would be health checking the system and removing services from the load balancer when they go down. 

It would also be used to add them to the load balancer when new containers come online. Is it time to start talking about implementing some type of service discovery agent to full this need?

**Comment by bacongobbler** [comment]
> Is it time to start talking about implementing some type of service discovery agent to fulfill this need?

That would be [publisher](http://docs.deis.io/en/latest/understanding_deis/components/#publisher). :)

Thank you for your comments. Let me come up with a proper response to answer both of your comments. I need to chew on this for a sec. :tea:

**Comment by bacongobbler** [comment]
> This also made me stop and think. Perhaps that should be the default, but the time should be configurable? And generally speaking I'm also wary of putting functionality in the controller if we can think of another way to do it.

Yep! I made the same comment [earlier in this thread](https://github.com/deis/deis/issues/4045#issuecomment-121767735).

> But I assume making the release promotion strategy configurable is out of scope for this particular change...

Yes. Those are nice changes, but it is out of scope.

> Normally your service discovery agent would be health checking the system and removing services from the load balancer when they go down.

I completely agree. However, I have left out the service discovery agent from performing health checks in this design for two reasons:

1) The controller would have no idea if the container is truly "up" or "down", making it difficult to give the app owner a good response about the app's health. However, we could expose an API endpoint from publisher to publish an app's health information. That could be a possible replacement to the entire `_do_healthcheck()` workflow in the controller. I'll keep chewing on this one.
2) publisher's workflow conflicts with [the controller's workflow](https://github.com/deis/deis/pull/3982#issuecomment-120112608) when we tried it out in an earlier iteration, however 1)'s idea by exposing an endpoint in publisher makes this moot. The problem with that workflow is yet another coupling between microservices, which may or may not be a bad thing since the controller is meant to be "the mainframe" for the entire platform (IOW as long as it's not the other way around it should be fine).

@jchauncey how would you imagine this would work, from your perspective?

**Comment by jchauncey** [comment]
i think in an idela world i would want this extracted into some external service that is responsible for the overall health of the cluster. It would perform both http and cmd health checks and perform some task based on that health (like adding/removing them from the load balancer and alerting). I think for the sake of time and knowing that the controller already performs several of the tasks you need I think having it there is probably the best compromise for now.

**Comment by technosophos** [comment]
I'm a little late to the discussion on this, but... that's never stopped me from weighing in before. ;-)

Kubernetes provides a good pattern for how we want to do health checks.

Containers can expose a healthcheck. There are two probe types that a container provides:

- Liveness: Whether the container is alive (failure = restart). Liveness is used to monitor health of existing pods.
- Readiness: Whether the container is ready to serve responses. Rediness is used to determine whether a pod's containers are all ready to handle requests.

Both probe types have the same data structure. K8s can do three types of probe:

- exec a command in the Docker contaner
- send an HTTP request and expect a 200 response
- make a TCP connection and expect the connection to be successfully established

I think we're mainly interested in HTTP probes right now, though I think exec probes will be useful for workers. An HTTP probe is pretty simple. Checks have the following major attributes:

- initialDelaySeconds: How long to wait before the first liveness probe
- timeOut: How long to wait for a response before marking unhealthy
- scheme, domain, port, path: To construct the URL to query

And that's about it. Here's an example usage from the User Guide:
``` yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-http
spec:
  containers:
  - args:
    - /server
    image: gcr.io/google_containers/liveness
    livenessProbe:
      httpGet:
        path: /healthz
        port: 8080
      initialDelaySeconds: 15
      timeoutSeconds: 1
    name: liveness
```
All HTTP probes are GETs (not configurable) and all expect a 200 status code (not configurable). I personally think this is fine. 

In Kubernetes, health checks are considered part of scheduling. Kublets execute health checks, and when a container fails a healthcheck, a replacement is spun up to take its place.

My opinion is that we should stick pretty close to this definition. While a methodology match would have us doing this as part of Publisher, I could see shunting this to another part of the platform purely for pragmatic reasons.

Workflow-wise, what we'd need is to handle two cases: Standard health polling and zero-downtime upgrades. Health polling is a relatively easy case. But what are the minimum number of changes we'd need to make to implement this style of health check?

## References:
* Pod health: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-guide/pod-states.md
* Description of Liveness Probes (exec and http): https://github.com/GoogleCloudPlatform/kubernetes/tree/master/docs/user-guide/liveness
* Probe type: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1/types.go#L625
* Liveliness example: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-guide/liveness/http-liveness.yaml

**Comment by bacongobbler** [comment]
Awesome! So to take in your feedback:

 - Remove HEALTHCHECK_STATUS_CODE from being configurable, enforce 200 OK
 - Keep HEALTHCHECK_URL to match `path`
 - Add HEALTHCHECK_INITIAL_DELAY (default = 0), which will have an initial delay of x seconds before performing the health check
 - Add HEALTHCHECK_TIMEOUT (default = 60), which will determine how long we will wait until we determine that the endpoint is unhealthy.

I'm not too sure how we can handle the `port` option with our current architecture, since the router listens on port 80 for all apps. Maybe we can set it to 80 by default and expose it as HEALTHCHECK_PORT? This would allow users who have https enabled in their routers to be able to hit their apps as well. How does that sound?

**Comment by jchauncey** [comment]
Shouldnt you hit the app directly including the port they expose (not 80).

**Comment by jchauncey** [comment]
like 
`http://<ip of machine here>:5000/healthcheck`

**Comment by bacongobbler** [comment]
I guess we could, but the controller at the present time does not have knowledge from the scheduler as to which host each instance of the app is running. However, we could try to extrapolate that information from the scheduler and then hit the app instance directly. In fleet's case, that would be the machine's private IP address. For kubernetes, that would likely be the DNS name.

For buildpack apps the user is not supposed to care which port is exposed, but I guess for this workflow we could just document that they should set HEALTHCHECK_PORT to 5000 and be done with it.

**Comment by jchauncey** [comment]
So was the thought to healthcheck through the router then?

**Comment by technosophos** [comment]
I think that initially we should not expose configuration for healthcheck port. If there's significant demand, then we can re-think that as an addition. This gives us the opportunity to enjoy a little extra flexibility in how we implement this first version.

**Comment by technosophos** [comment]
Also, one thing I didn't make too clear in my earlier post: I think that liveness and readiness checks can also be conflated for the first release(s). I don't think we need to give separate options for both. I think @bacongobbler is assuming that already.

**Comment by jchauncey** [comment]
You cant health check an app through the router because you need to know which container is down/up when performing actions against the load balancer. This means you need to know which port the application is running on. So you can either determine that programmatically or you can have them specify it as part of some config detail when creating an application.

**Comment by technosophos** [comment]
Yeah, I don't think we should rely on the user to set the port.

**Comment by bacongobbler** [comment]
> So was the thought to healthcheck through the router then?

Yes, the initial thought was to do the healthcheck through the router. If we want to hit each instance directly, how would the readiness checks in the controller look? Would we pick one of the new instances at random and do a health check, or do a health check on each new instance?

> Also, one thing I didn't make too clear in my earlier post: I think that liveness and readiness checks can also be conflated for the first release(s). I don't think we need to give separate options for both. I think @bacongobbler is assuming that already.

Yes, I am assuming that at the moment since we don't have something that acts as a Liveness checker. This discussion is mostly wrapped around emulating a Readiness checker :)

> This means you need to know which port the application is running on. So you can either determine that programmatically or you can have them specify it as part of some config detail when creating an application.

> Yeah, I don't think we should rely on the user to set the port.

OTOH, the user would know which port the app is expecting to expose because it's typically [right in the Docker image](http://docs.docker.com/reference/builder/#expose). For users who use the Heroku workflow, we can document that this should be port 5000 as specified [here](https://github.com/deis/deis/blob/master/builder/rootfs/usr/local/src/slugrunner/Dockerfile#L12). Alternatively, we can use the Docker API to do this programatically, but then the user won't have a way to say which port to use if the app exposes more than one port.

**Comment by jchauncey** [comment]
As the controller launches new instances it should also check those instances first before they ever get added to the load balancer. if an application is flapping you cannot place that container in the rotation. 

The controller should just do a simple http check against the ip of the host and the port that it exposes as its web port (which i thought was not necessarily configurable but not always 5000). 

Like this - https://github.com/RallySoftware/armada/blob/master/lib/armada/connection/health_check.rb

**Comment by technosophos** [comment]
@bacongobbler said:
> OTOH, the user would know which port the app is expecting to expose because it's typically right in the Docker image. For users who use the Heroku workflow, we can document that this should be port 5000 as specified here. Alternatively, we can use the Docker API to do this programatically, but then the user won't have a way to say which port to use if the app exposes more than one port.

Alright, I'm convinced. We can do a HEALTHCHECK_PORT.

**Comment by technosophos** [comment]
@smothiki and @kmala -- anything you'd like to add to this discussion?

**Comment by bacongobbler** [comment]
I have updated the OP to address changes discussed in this conversation. When you have time, PTAL, otherwise I'll consider this document as final and start working on an implementation :)

**Comment by Blystad** [comment]
Hmm, so, if I understand correctly, you will first do a healthcheck at `HEALTHCHECK_INITIAL_DELAY` instead of attempting for up to 60 (configurable) seconds. Now, if, after `HEALTHCHECK_INITIAL_DELAY` the healthcheck fails, then the publish process will be cancelled, right?

With this solution, a user would probably have to configure `HEALTHCHECK_INITIAL_DELAY` at somewhere between 150% and 200% of the actual expected time for the first healthcheck. E.g. if I expect my application to be up and running in 10 seconds, I would configure this probably at like 20 seconds. (In case of new code slowing it down, or database migrations, or other stuff taking extra time). 

This would cause every new deploy to take extra time (20 seconds instead of 10 seconds), which is not a lot, but over time it does add up (+ it's annoying). 

OR, do you mean that you will first wait `HEALTHCHECK_INITIAL_DELAY` and then keep trying for `HEALTHCHECK_TIMEOUT`? 

E.g. if I configure `HEALTHCHECK_INITIAL_DELAY` to 10 seconds (as I expect the application to start in around 10 seconds), and then leave `HEALTHCHECK_TIMEOUT` at it's default 60s, then it will try to establish a connection to `HEALTHCHECK_URL` with 1s connect/data timeout for 60s?

**Comment by bacongobbler** [comment]
Your first assumption is correct. Take a look at @mbutcher's comment above for context :)

**Comment by jchauncey** [comment]
Wait so you will fail the publish if the first check fails?

**Comment by bacongobbler** [comment]
Yes. The first request after the initial delay should give us the correct response (or a failboat)

**Comment by jchauncey** [comment]
I think thats the wrong way to do this. Ive never seen a health check
system that fails after the first try. All apps have a variable time in
starting up. Sometimes it could take 60 seconds the next 65, especially
complicated java apps that are loading a ton of spring beans.

I think you want to have a retry_delay (number of seconds between tries)
and then a max number of tries. Both should be configurable.

If we choose the proposed solution I know of a number of apps that would
fail every time they were deployed because of the variability in startup
times.

**Comment by mackwic** [comment]
I second @jchauncey . It's an easy trick which simplify the implementation but degrade the dev experience.

Also I didn't understand why the heathchecker should knock only once for startup. Why not keeping it for the whole life of the node, reporting healthiness for all running services ?

It can be something simple inspired from the gossip protocols: each second, each life-checker of each node chose a random service from the whole cluster, and check its liveliness. It then update the correct key in etcd with last time update and status. (it's the way [Serf](serfdom.io) works with membership).

We need to run the numbers to see when statistically we would hit a new service considering different coreos configurations. But a prioritization can also be done: the new-service list is prioritized against the list of already running services. This way we know sooner when they are up without degrading the global health-checking too much.

Last but not least, it would be nice to have more than one healthcheck point. The configuration can be trivial: the `GET` request on the healthcheck endpoint _can_ return one or more `Location:` headers, which would then be checked. That one would be neat.

**Comment by Blystad** [comment]
I as well agree with @jchauncey . 

Secondly, another thing, the `/up` endpoint:

```
GET /up
```
which returns this list:
```
[
    "foo_v2.web.1",
    "foo_v2.web.3",
    "bar_v3.cmd.1"
]
```

Wouldn't it be better to filter this, at least by application, and possibly by version?

E.g.

```
GET /up/foo
GET /up/bar
```

and/or

```
GET /up/foo/v2
GET /up/foo/v3
GET /up/bar/v2
```

I don't see the point of A) sending a list of every living container to the controller, and B) Have the controller iterate through this list.

**Comment by technosophos** [comment]
So the ideas of liveness and readiness have gotten confused again.

*Readiness* should check for when a container is ready to serve.
*Liveness* checks whether a container which was previously marked "ready" is still alive.

One healthz endpoint can be used for both. But the logic behind checking for readiness and checking for liveness are different. Readiness is checked only during container deployment, and not until the initial timeout has passed. Liveness is checked repeatedly over the lifetime of the app.

Kubernetes does 3 retries of each type of probe (Liveness or readiness). I don't think they insert delays between probes, but they might. I think we should probably try an incremental backoff strategy that starts with 10% of the HEALTCHECK_INITIAL_DELAY, then 50%, then 100%.

The max retries is not configurable, and I don't actually see any compelling reason to make it so for Deis -- at least not for the Deis user, who could do something horrendous like set it to a million.

(For Kubernetes, see https://github.com/GoogleCloudPlatform/kubernetes/blob/0048eae8ebd4311f66de786f24ae56ec97bc0040/pkg/kubelet/prober/prober.go)

As far as the container developer is concerned, only one healthz endpoint needs to be implemented. Readiness and liveness can both be checked on that same URL. So nothing in this model makes life more complex for the Deis user.

**Comment by mackwic** [comment]
> So the ideas of liveness and readiness have gotten confused again.

The state of healthy/not-healthy is _less_ subtle than ready/alive.
What I suggest, is to separate the logic of testing "is it healthy ?", store this info somewhere, and it's the job to someone else to understand what it means (ready/not-ready or alive/not-alive).

What do you think ?

**Comment by bacongobbler** [comment]
So to explain this more clearly (thanks @technosophos), the health checker indicates whether the container is ready to service requests. If it fails, publisher will not publish it to etcd. The default state before the initial delay is `Down`. The state for a container when no health check is provided is assumed to be `Up` (assuming the container is actually "up"). This will run forever, continuously checking if the container is up or down, publishing and unpublishing as needed. This is the *Liveness* check.

Now, here's where the controller comes into play: the controller is the unit that deploys and recycles containers as needed, since it manages an application's lifecycle. Because of this, it needs to perform a *Readiness* check after an initial delay. For Heroku, this is at 60 seconds, which then if the container is not listening on the expected port then the dyno is killed off. For us, we need to check to ensure the new release is up and responding. If it is down, then we need to kill it and keep the old containers.

> I think we should probably try an incremental backoff strategy that starts with 10% of the HEALTHCHECK_INITIAL_DELAY, then 50%, then 100%.

That sounds like a happy medium for the liveness checker.

> Wouldn't it be better to filter this, at least by application, and possibly by version?

Sure. The idea of a global list is to show all containers being published on a single host (since publisher works on a host-by-host basis), but I can add query filtering.

**Comment by bacongobbler** [comment]
@mackwic I think you're getting confused as that's what is happening here. Publisher is the component that tests "is it healthy?" and the controller is the one that determines the state of the application's containers.

**Comment by jchauncey** [comment]
For a 60 second intial delay I would wait again for 6 seconds, then 30 seconds then 60 seconds? So the last check would happen 1 minute and 36 seconds after the initial check (60 seconds).

**Comment by bacongobbler** [comment]
Yes, that would the theoretical maximum your app has to respond before your app is considered kaput.

**Comment by jchauncey** [comment]
I guess that means your application could be up but your still waiting. If you app comes up in 97 seconds then you have to wait 59 seconds before its added into the rotation. Still think it should be configurable but know that brings in more trouble for you guys because you need to store that info somewhere.

**Comment by mbutcher** [comment]
Just an FYI… I’m not the @mbutcher you are looking for. *chuckle* How did I get on this project?

**Comment by Joshua-Anderson** [comment]
@mbutcher: @technosophos is mbutcher elsewhere. I'm just happy I haven't made that mistake yet :smiley:.

**Comment by technosophos** [comment]
@jchauncey that is how incremental backoff works. If you're hitting the last of the backoff intervals, you really need to tune your `HEALTHCHECK_INITIAL_DELAY`. But in the meantime, the app isn't getting slammed with healthcheck requests.

We don't *have* to do it that way, but it is an effort to tolerate the 250% case while not doing tons of healthchecks.

**Comment by jchauncey** [comment]
but wont you be slamming hte same endpoint anyways when you do liveness checks?

**Comment by technosophos** [comment]
The assumption with a liveness test is that the container is already in a stable state (while startup is not a stable state).

@bacongobbler Do you think we should set max values on `HEALTHCHECK_INITIAL_DELAY` and `HEALTHCHECK_TIMEOUT`? Timeouts in particular consume network connections, so making sure nobody sets theirs to an extraordinarily high value might be nice.

**Comment by bacongobbler** [comment]
That might be a good idea. 20 minutes would [match the rest](https://github.com/deis/deis/blob/master/controller/scheduler/fleet.py#L191) of our [timeout values](https://github.com/deis/deis/blob/master/controller/templates/gconf.py#L11), so that would be a good maximum.

**Comment by bacongobbler** [comment]
Any other thoughts on this?

**Comment by mboersma** [comment]
I've read this a few times and I'm happy with this proposal and the great discussion around it. So long as we don't intentionally code anything that makes working more directly with k8s difficult, I'd say let's not let perfect be the enemy of good. :+1:

**Comment by bacongobbler** [comment]
I think there's one thing wrong with this proposal and it's the idea of the controller hitting publisher. publisher is an agent running on every host, so if we wanted to perform the readiness check, it would have to every host, aggregate the information and then determine if the app is running. I'm thinking that it'd be easier to just connect to etcd and check the keyspace to see if the new version of the app is running instead.
